Sources/Benchmark/BenchmarkSetting.swift

```diff
@@ -12,68 +12,106 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-public enum BenchmarkSetting {
-    case iterations(Int)
-    case maxIterations(Int)
-    case warmupIterations(Int)
-    case filter(String)
-    case minTime(seconds: Double)
+public protocol BenchmarkSetting {}
+
+public struct Iterations: BenchmarkSetting {
+    public var value: Int
+    public init(_ value: Int) {
+        self.value = value
+    }
 }
 
-struct BenchmarkSettings {
-    let iterations: Int?
-    let warmupIterations: Int?
-    let maxIterations: Int
-    let filter: BenchmarkFilter
-    let minTime: Double
+public struct MaxIterations: BenchmarkSetting {
+    public var value: Int
+    public init(_ value: Int) {
+        self.value = value
+    }
+}
 
-    init(_ settings: [[BenchmarkSetting]]) throws {
-        try self.init(Array(settings.joined()))
+public struct WarmupIterations: BenchmarkSetting {
+    public var value: Int
+    public init(_ value: Int) {
+        self.value = value
     }
+}
 
-    init(_ settings: [BenchmarkSetting]) throws {
-        var iterations: Int? = nil
-        var warmupIterations: Int? = nil
-        var maxIterations: Int = -1
-        var filter: String? = nil
-        var minTime: Double = -1
+public struct Filter: BenchmarkSetting {
+    public var value: String
+    public init(_ value: String) {
+        self.value = value
+    }
+}
+
+public struct MinTime: BenchmarkSetting {
+    public var value: Double
+    public init(seconds value: Double) {
+        self.value = value
+    }
+}
+
+public struct BenchmarkSettings {
+    let settings: [String: Any]
+
+    public init(_ settings: [[BenchmarkSetting]]) {
+        self.init(Array(settings.joined()))
+    }
+
+    public init(_ settings: [BenchmarkSetting]) {
+        var result: [String: BenchmarkSetting] = [:]
 
         for setting in settings {
-            switch setting {
-            case .iterations(let value):
-                iterations = value
-            case .warmupIterations(let value):
-                warmupIterations = value
-            case .filter(let value):
-                filter = value
-            case .maxIterations(let value):
-                maxIterations = value
-            case .minTime(let value):
-                minTime = value
+            let key = String(describing: type(of: setting))
+            result[key] = setting
+        }
+
+        self.init(result)
+    }
+
+    public init() {
+        self.init([:])
+    }
+
+    init(_ settings: [String: BenchmarkSetting]) {
```

Comment:
We type name because `Any.Type` is not Hashable by default. If we had support for private/scoped conformances we could define one that uses type name, but otherwise it seems risky to provide it globally.